File: LeetCode/54. Spiral Matrix.py
# 思路是比较简单的，模拟他的前进轨迹，判断好边界条件即可
# 看了下答案，这个写法还可以优化，把从(r1,c1)到(r1,c2)封装成函数，并且方向用数组表示，可以省略很大一部分代码

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        row_len = len(matrix)
        if (row_len == 0):
            return  []
        col_len = len(matrix[0])
        if (col_len == 0):
            return []
        times = row_len * col_len
        side = 0
        count = 0
        result = []
        row = col = 0
        round_count = 0
        while (count < times):
            result.append(matrix[row][col])
            if (side == 0):
                if (col < col_len - 1 - round_count):
                    col += 1
                else:
                    row += 1
                    side = 1
            elif (side == 1):
                if (row < row_len - 1 - round_count):
                    row += 1
                else:
                    col -= 1
                    side = 2
                    
            elif (side == 2):
                if (col > round_count):
                    col -= 1
                else:
                    row -= 1
                    side = 3
            else:
                if (row > round_count + 1):
                    row -= 1
                else:
                    col += 1
                    side = 0
                    round_count += 1
                
            count += 1
            
        return result

File: LeetCode/54. Spiral Matrix.cpp
// 这道题麻烦的主要是细节，要想清楚边界条件，思路很简单，复杂度也是一定的

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans = vector<int>();
        int row_size = matrix.size();
        if (row_size == 0) {
            return ans;
        }
        int col_size = matrix[0].size();
        int row = 0;
        int col = 0;
        int row_min = -1;
        int row_max = row_size;
        int col_min = -1;
        int col_max = col_size;
        int dir[4][2] = {
            {0, 1},
            {1, 0},
            {0, -1},
            {-1, 0},
        };
        int dir_index = 0;
        while(row > row_min && row < row_max && col > col_min && col < col_max) {
            ans.push_back(matrix[row][col]);
            if (row + dir[dir_index][0] <= row_min) {
                col_min += 1;
                dir_index = (dir_index + 1) % 4;
            }
            else if (row + dir[dir_index][0] >= row_max) {
                col_max -= 1;
                dir_index = (dir_index + 1) % 4;
            }
            else if (col + dir[dir_index][1] <= col_min) {
                row_max -= 1;
                dir_index = (dir_index + 1) % 4;
            }
            else if (col + dir[dir_index][1] >= col_max) {
                row_min += 1;
                dir_index = (dir_index + 1) % 4;
            }

            row = row + dir[dir_index][0];
            col = col + dir[dir_index][1];
        }

        return ans;
    }
};